### ISSUE TITLE
CapabilityBasedPartitioner doesn't support horizontal (vertical?) fusion

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#81531]
### 🐛 Describe the bug

steps to reproduce:

```
diff --git a/test/test_fx_passes.py b/test/test_fx_passes.py
index 39fe501f44..9fe5e4002e 100644
--- a/test/test_fx_passes.py
+++ b/test/test_fx_passes.py
@@ -158,6 +158,12 @@ class TestPartitionFunctions:
         out = torch.stack([add_1, add_2, add_3])
         return out
 
+    @staticmethod
+    def forward12(a):
+        add_1 = a + 1
+        add_2 = a + 1
+        return add_1, add_2
+
 # A mock OperatorSupport class, where only operator.add is supported
 class MockOperatorSupport(OperatorSupport):
     def is_node_supported(self, submodules, node: torch.fx.Node) -> bool:
@@ -179,6 +185,7 @@ class TestFXGraphPasses(JitTestCase):
         (TestPartitionFunctions.forward9, [['add_3', 'add_2', 'add_1', 'add']]),
         (TestPartitionFunctions.forward10, [['add_3', 'add_2', 'add', 'add_1']]),
         (TestPartitionFunctions.forward11, [['add_1'], ['add']]),
+        (TestPartitionFunctions.forward12, [['add_1', 'add_2']]),
     ])
     def test_partitioner(self, fn, expected_partition):
         traced = symbolic_trace(fn)
@@ -188,7 +195,7 @@ class TestFXGraphPasses(JitTestCase):
         partitions = partitioner.propose_partitions()
 
         partitions_name = [[node.name for node in partition.nodes] for partition in partitions]
-        assert len(partitions_name) == len(expected_partition)
+        assert len(partitions_name) == len(expected_partition), partitions_name
         for i in range(len(partitions_name)):
             assert set(partitions_name[i]) == set(expected_partition[i])
 
```

I don't know if nvFuser supports horizontal fusion or not, but if the backend supports it (e.g., cuda graphs does) loss of fusion here is bad as you are doubling your memory traffic.

cc @SherlockNoMad @Chillee 

### Versions

master
